02_rand_dataset.ipynb

```diff
@@ -33,56 +24,400 @@
+    "def read_file_to_list(file_name):\n",
     "    with open(file_name, 'r') as file:\n",
-    "        return file.read().decode('utf-8')\n",
+    "        return file.readlines()\n",
+    "\n",
+    "# The file has been re-encoded in UTF-8, the source encoding is Latin-1\n",
+    "terrorism_articles = read_file_to_list('data/rand-terrorism-dataset.txt')\n",
+    "\n",
+    "# Create a list of spaCy Doc objects representing articles\n",
+    "terrorism_articles_nlp = [nlp(art) for art in terrorism_articles]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example output using the full ~40,000 article dataset\n",
+    "\n",
+    "![Heatmap of terrorist group and country](images/example_output_full.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Example solution"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define some geographical areas and groups to inspect\n",
+    "\n",
+    "These are commonly mentioned in the full dataset, which you can prove for yourself using the same approaches as in the previous tutorial for Pride and Prejudice. You can process the full dataset at once into one spaCy span using the read_file() function. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 69,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "common_terr = ['taliban', \n",
+    "               'al - qaeda', \n",
+    "               'hamas',  \n",
+    "               'fatah', \n",
+    "               'plo', \n",
+    "               'bilad al - rafidayn'\n",
+    "              ]\n",
+    "\n",
+    "common_locs = ['iraq',\n",
+    "               'baghdad', \n",
+    "               'kirkuk', \n",
+    "               'mosul', \n",
+    "               'afghanistan', \n",
+    "               'kabul',\n",
+    "               'basra', \n",
+    "               'palestine', \n",
+    "               'gaza', \n",
+    "               'israel', \n",
+    "               'istanbul', \n",
+    "               'beirut', \n",
+    "               'pakistan'\n",
+    "              ]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Inspect each article for mentions of groups and locations"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 70,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "11"
+      ]
+     },
+     "execution_count": 70,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "location_entity_dict = defaultdict(Counter)\n",
+    "\n",
+    "for article in terrorism_articles_nlp:\n",
+    "    #Get all the groups and locations in the article\n",
+    "    article_ents = [ent.lemma_ for ent in article.ents if ent.label_ == 'PERSON' or ent.label_ == 'ORG']\n",
```

Comment:
I think variable name `ents` is a bit misleading, since you are storing terrorist candidates. I would go with something like `article_terrorist_candidates`.